### ISSUE TITLE
Return type hint can be used: static is suggested

### ISSUE THREAD

**Issue by bazzik** [kalessil/phpinspectionsea#730]
PhpStorm 2017.3.2; plugin version: 2.3.15.1; PHP Language Level is set to 7.0

Got new suggestions after upgrading this plugin today. Generates wrong code that PhpStorm complains about straight away.

```
<?php
class SomeClass
{
    protected $someVar;
    
    /**
     * Some clever method
     * 
     * @return $this
     */
    public function doSomething()
    {
        $this->doThat();
        $this->someVar++;

        return $this;
    }

    protected function doThat()
    {}
}
```
As of today the inspection suggest me that 
> ': static' can be declared as return type hint

![screen01](https://user-images.githubusercontent.com/1482298/34586100-83eb7488-f19a-11e7-885e-0ac0a6484ee8.png)

If I use proposed quick fix it generates the wrong code that PhpStorm complains straight way:
```
    public function doSomething(): static
```
![screen02](https://user-images.githubusercontent.com/1482298/34586084-7003e7ac-f19a-11e7-8458-ac34e167e7e6.png)

**Comment by tomzx** [comment]
@kalessil It seems that https://github.com/kalessil/phpinspectionsea/blob/C-2.3.15.1/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/ReturnTypeCanBeDeclaredInspector.java#L53 shouldn't be there as it is not a valid return type.

**Comment by kalessil** [comment]
That's some magic inside IDE types resolving, seems it resolved $this as static and I need fix type hint generation.

I'll make a fix, the short-term solution is to disable an option saying to respect return-annotation.

**Comment by bazzik** [comment]
>That's some magic inside IDE types resolving, seems it resolved $this as static

Pretty much -- I can use it instead `$this` in `@return` tags and the effect will be the same in terms of code completion behavior regardless if it's a static or normal method.

**Comment by Logiar** [comment]
Since it seems related I'll mention it can suggest : void as a return type for getters, like in this code:
```php
<?php

class TestClass
{
    protected $something;

    public function mayChangeIt(): void
    {
        $this->something = 'Now has a value';
    }

    public function getSomething()
    {
        return $this->something;
    }
}

```

**Comment by kalessil** [comment]
@Logiar : more or less related, I'll check what why the inspection doesn't suggest `void` there.

**Comment by kalessil** [comment]
@Logiar : I see now, thank you for clarification.

**Comment by kalessil** [comment]
@bazzik : fixed
@Logiar : fixed

**Comment by bazzik** [comment]
@kalessil 
May I ask how it was fixed?
I was kind of expecting to see ": self" offered instead of invalid ": static" .. but inspection does not offer anything.

If I change `@return $this` to be `@return self` then inspection offers adding ": self".

Am I misunderstanding how it meant to work (actual PHP code)? It seem to wok OK here:
```
<?php
declare(strict_types=1);

class SomeClass
{
    protected $someVar;

    /**
     * @return $this
     */
    public function doSomething(): self
    {
        $this->doThat();
        $this->someVar++;

        return $this;
    }

    protected function doThat()
    {}
}

class AnotherClass extends SomeClass {
    public function sayYes(): self
    {
        //TODO: Implement sayYes() method
        echo "Yes!\n";

        return $this;
    }
}

(new AnotherClass())->doSomething()->sayYes()->sayYes();
```

**Comment by kalessil** [comment]
@bazzik : I added logic filtering `$this` and `static` in PhpDoc when suggesting `self` as return type.

I can translate `$this` and `static` into `self` - will it be correct?

**Comment by bazzik** [comment]
@kalessil 
Based on current PHP 7.1/Laravel 5.5 project + plus taking that code sample above I say: offer `: self` if you see `@return $this` -- it seem to work OK. Maybe introduce a check box for that inspection if desired (but make it enabled by default).

I would leave `@return static` cases aside for the moment (as it really meant to be used in static context only ... despite the fact that it works the same as `@return $this` in PhpStorm (at least in all of my experiments in this regard)).

**Comment by kalessil** [comment]
@bazzik : thank you for checking, we are good to go with #761 then (`@return $this` -> `self`)
